File: src/2024/day-01/python/main.py
from part_1 import Part1
from part_2 import Part2

input = open('./data/input.txt', 'r', encoding='utf-8').read()
test_input_1 = open('./data/test-input-1.txt', 'r', encoding='utf-8').read()
test_input_2 = open('./data/test-input-2.txt', 'r', encoding='utf-8').read()

if __name__ == "__main__":
    print("---- Part 1 ----")
    test = Part1(test_input_1)
    print(f" Test:  {test}")
    final = Part1(input)
    print(f" Final: {final}")
    print("---- Part 2 ----")
    test = Part2(test_input_2)
    print(f" Test:  {test}")
    final = Part2(input)
    print(f" Final: {final}")

File: src/2024/day-01/rust/src/main.rs
use std::time::Instant;
mod part_1;
mod part_2;

fn main() {
  // input
  const INPUT: &str = include_str!("../data/input.txt");
  const TEST_INPUT_1: &str = include_str!("../data/test-input-1.txt");
  const TEST_INPUT_2: &str = include_str!("../data/test-input-2.txt");

  println!("================== Day 01 ==================");

  // part 1
  println!("------------------ Part 1 ------------------");
  let start = Instant::now();
  let result = part_1::part_1(TEST_INPUT_1);
  let duration = start.elapsed();
  println!("Test:   {}", check_result(result, 142, duration));

  let start = Instant::now();
  let result = part_1::part_1(INPUT);
  let duration = start.elapsed();
  println!("Answer: {}", check_result(result, 54667, duration));

  // part 2
  println!("------------------ Part 2 ------------------");
  let start = Instant::now();
  let result = part_2::part_2(TEST_INPUT_2);
  let duration = start.elapsed();
  println!("Test:   {}", check_result(result, 281, duration));

  let start = Instant::now();
  let result = part_2::part_2(INPUT);
  let duration = start.elapsed();
  println!("Answer: {}", check_result(result, 54203, duration));
}

fn check_result<T: std::fmt::Display + std::cmp::PartialEq>(result: T, expected: T, duration: std::time::Duration) -> String {
  if result == expected {
    format!("{:8}, pass ✅, time: {:?}", result, duration)
  } else {
    format!("{:8}, fail ❌, expected {}", result, expected)
  }
}
